perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3057,56 +3058,47 @@ private void updateExistingMembersWhileSynchronization(PerunSession sess, Group
-						if (currentMembershipExpirationDate.before(now)) {
-							//disabled members which are after expiration date will be expired
-							if (richMember.getStatus().equals(Status.DISABLED)) {
-								try {
-									perunBl.getMembersManagerBl().expireMember(sess, richMember);
-									log.info("Switching member id {} to EXPIRE state, due to expiration {}.", richMember.getId(), membershipExpiration.getValue());
-									log.debug("Switching member to EXPIRE state, additional info: membership expiration date='{}', system now date='{}'", currentMembershipExpirationDate, now);
-								} catch (MemberNotValidYetException e) {
-									log.error("Consistency error while trying to expire member id {}, exception {}", richMember.getId(), e);
-								}
-							}
-						} else {
-							//disabled and expired members which are before expiration date will be validated
-							try {
-								perunBl.getMembersManagerBl().validateMember(sess, richMember);
-								log.info("Switching member id {} to VALID state, due to expiration {}.", richMember.getId(), membershipExpiration.getValue());
-								log.debug("Switching member to VALID state, additional info: membership expiration date='{}', system now date='{}'", currentMembershipExpirationDate, now);
-							} catch (WrongAttributeValueException | WrongReferenceAttributeValueException e) {
-								log.error("Error during validating member id {}, exception {}", richMember.getId(), e);
-							}
-						}
+						currentMembershipExpirationDate = BeansUtils.getDateFormatterWithoutTime().parse((String) membershipExpiration.getValue());
+						memberExpiredInPast = currentMembershipExpirationDate.before(now);
 					} catch (ParseException ex) {
-						log.error("Group synchronization: memberId {} expiration String cannot be parsed, exception {}.",richMember.getId(), ex);
+						log.error("Group synchronization: memberId {} expiration String cannot be parsed, exception {}.", richMember.getId(), ex);
 					}
 				}
-			}
-
-			// If the member has INVALID status, try to validate the member
-			try {
-				if (richMember.getStatus().equals(Status.INVALID)) {
-					getPerunBl().getMembersManagerBl().validateMember(sess, richMember);
-				}
-			} catch (WrongAttributeValueException | WrongReferenceAttributeValueException e) {
-				log.info("Member id {} will stay in INVALID state, because there was problem with attributes {}.", richMember.getId(), e);
-			}
 
-			// If the member has still DISABLED status, try to validate the member
-			try {
-				if (richMember.getStatus().equals(Status.DISABLED)) {
-					getPerunBl().getMembersManagerBl().validateMember(sess, richMember);
+				if ((Status.DISABLED.equals(memberStatus) || Status.INVALID.equals(memberStatus)) && memberHasExpiration && memberExpiredInPast) {
+					//If member has expiration in the past (should be expired now) and is in other state than expired, expire him
+					try {
+						perunBl.getMembersManagerBl().expireMember(sess, richMember);
```

Comment:
```suggestion
						richMember = (RichMember) perunBl.getMembersManagerBl().expireMember(sess, richMember);
```
expireMember method change status of object as a side effect. So either put it into documentation of expireMember or rather not reply on it.

Reply:
Added to javadoc an information about side effect.